src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -257,11 +261,146 @@ impl Coordinator {
         })
     }
 
+    /// Subsources are planned differently from other statements because they
+    /// are typically synthesized from other statements, e.g. `CREATE SOURCE`.
+    /// Because of this, we have usually "missed" the opportunity to plan them
+    /// through the normal statement execution life cycle (the exception being
+    /// during bootstrapping).
+    pub(crate) async fn plan_subsource(
+        &mut self,
+        session: &Session,
+        params: &mz_sql::plan::Params,
+        subsource_stmt: CreateSubsourceStatement<mz_sql::names::Aug>,
+    ) -> Result<CreateSourcePlanBundle, AdapterError> {
+        let resolved_ids = mz_sql::names::visit_dependencies(&subsource_stmt);
+        let source_id = self.catalog_mut().allocate_user_id().await?;
+        let plan = self.plan_statement(
+            session,
+            Statement::CreateSubsource(subsource_stmt),
+            params,
+            &resolved_ids,
+        )?;
+        let plan = match plan {
+            Plan::CreateSource(plan) => plan,
+            _ => unreachable!(),
+        };
+        Ok(CreateSourcePlanBundle {
+            source_id,
+            plan,
+            resolved_ids,
+        })
+    }
+
+    /// Prepares an `ALTER SOURCE...ADD SUBSOURCE`.
+    pub(crate) async fn plan_purified_alter_source_add_subsource(
+        &mut self,
+        session: &Session,
+        params: Params,
+        id: GlobalId,
+        options: Vec<AlterSourceAddSubsourceOption<Aug>>,
+        create_subsource_stmts: Vec<CreateSubsourceStatement<Aug>>,
+    ) -> Result<(Plan, ResolvedIds), AdapterError> {
+        let mut subsources = Vec::with_capacity(create_subsource_stmts.len());
+        for subsource_stmt in create_subsource_stmts {
+            let s = self
+                .plan_subsource(session, &params, subsource_stmt)
+                .await?;
+            subsources.push(s);
+        }
+
+        let action = mz_sql::plan::AlterSourceAction::AddSubsourceExports {
+            subsources,
+            options,
+        };
+
+        Ok((
+            Plan::AlterSource(mz_sql::plan::AlterSourcePlan { id, action }),
+            ResolvedIds(BTreeSet::new()),
+        ))
+    }
+
+    /// Prepares a `CREATE SOURCE` statement to create its progress subsource,
+    /// the primary source, and any data-bearing subsources (e.g. PG tables).
+    pub(crate) async fn plan_purified_create_source(
+        &mut self,
+        ctx: &ExecuteContext,
+        params: Params,
+        progress_stmt: CreateSubsourceStatement<Aug>,
+        mut source_stmt: mz_sql::ast::CreateSourceStatement<Aug>,
+        subsource_stmts: Vec<CreateSubsourceStatement<Aug>>,
+    ) -> Result<(Plan, ResolvedIds), AdapterError> {
+        let mut create_source_plans = Vec::with_capacity(subsource_stmts.len() + 2);
+
+        // 1. First plan the progress subsource
+        // Progress subsource must not have a dependency on the source.
```

Comment:
This is why I dislike comments––all they do is get people asking hard questions. Here's my amendment:

```
// 1. First plan the progress subsource.
//
// The primary source depends on this subsource because the primary
// source needs to know its shard ID, and the easiest way of
// guaranteeing that the shard ID is discoverable is to create this
// collection first.
```